pilot/pkg/proxy/envoy/xds/xds.go

```diff
@@ -139,3 +161,48 @@ func (s *Server) StartGRPC(addr string) error {
 	}()
 	return nil
 }
+
+// ProxyGen implements a proxy generator - any request is forwarded using the agent ADSC connection.
+// Responses are forwarded back on the connection that they are received.
+type ProxyGen struct {
+	adsc   *adsc.ADSC
+	server *Server
+}
+
+// HandleResponse will dispatch a response from a federated
+// XDS server to all connections listening for that type.
+func (p *ProxyGen) HandleResponse(con *adsc.ADSC, res *discovery.DiscoveryResponse) {
+	// TODO: filter the push to only connections that
+	// match a filter.
+	p.server.DiscoveryServer.PushAll(res)
+}
+
+func (s *Server) NewProxy() *ProxyGen {
+	return &ProxyGen{
+		server: s,
+	}
+}
+
+func (p *ProxyGen) AddClient(adsc *adsc.ADSC) {
+	p.server.m.Lock()
+	p.adsc = adsc // TODO: list
```

Comment:
I am pretty concerned that our test code, which has multiple bugs/performance issues is suddenly the prod code (if this becomes default). I was expecting a more direct proxy like:

```go
for {
   switch {
      case xdsReq <- envoyStream.Recv():
         istiodStream.Send(xdsReq)
      case xdsResp <- istiodStream.Recv()
         envoyStream.Send(xdsResp)
   }
}
```

With relevant modifications added in there as needed of course, like envoyfilter patch we just modify `xdsResp` before we send it.

Have you considered this approach? Any pros/cons compared to this one?